File: src/db/types/requests.py
# pylint: disable=invalid-name
from __future__ import annotations
from typing import Annotated
from pydantic import model_validator, Field

from utils.schemas import BaseModel
from .validators import is_email


class RQ:
    class auth:
        class signin(BaseModel):
            id: int | None = None
            domain: str | None = None
            email: str | None = None
            password: str

            @model_validator(mode="after")
            def check_only_one_field(self):
                error_message = "You must provide exactly one of the fields: email, domain, user_id"
                assert sum(x is not None for x in (self.id, self.domain, self.email)) == 1, error_message
                return self

    class organizations:
        class get(BaseModel):
            id: int

        class get_groups(BaseModel):
            id: int

    class users:
        class get_organizations(BaseModel):
            id: int

    class groups:
        class get(BaseModel):
            id: int

    class test:
        class signup(BaseModel):
            first_name: str = Field(..., max_length=30)
            email: Annotated[str, is_email]
            password: str

        class organizations:
            class create(BaseModel):
                name: str = Field(..., max_length=30)

            class invite(BaseModel):
                organization_id: int
                user_id: int

        class groups:

            class create(BaseModel):
                name: str = Field(..., max_length=30)
                organization_id: int

            class invite(BaseModel):
                user_id: int
                group_id: int

File: tests/src/helpers/requests.ts
namespace RQ {
    export namespace auth {
        export interface signin {
            email?: string;
            domain?: string;
            id?: number;
            password: string;
        }
    }
    export namespace organizations {
        export interface get {
            id: number;
        }

        export interface get_groups {
            id: number;
        }
    }
    export namespace users {
        export interface get_organizations {
            id: number;
        }
    }
    export namespace groups {
        export interface get {
            id: number;
        }
    }
    export namespace test {
        export interface signup {
            firstName: string;
            email: string;
            password: string;
        }
        export namespace organizations {
            export interface create {
                name: string;
            }
            export interface invite {
                organizationId: number;
                userId: number
            }
        }
        export namespace groups {
            export interface create {
                name: string;
                organizationId: number;
            }
            export interface invite {
                userId: number;
                groupId: number;
            }
        }
    }
}

export default RQ;
